### ISSUE TITLE
.NET core 2.1.x , 2.2.x  wildcard certificate handling issue

### ISSUE THREAD

**Issue by madhub** [dotnet/corefx#34061]
I have self hosted asp.net web api with wild card certificate ( ***.mdrinc.com**)  running on port 44321.

When I try to reach server using HttpClient I get "_The remote certificate is invalid according to the validation procedure_" error . With HttpClient handler with ServerCertificateCustomValidationCallback I see "_SslPolicyErrors.RemoteCertificateNameMismatch_". 

Same program works with AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false); 

**Curl also works**
```
 curl -v https://ms.mdrinc.com:44321/api/values
*   Trying 161.85.18.231...
* TCP_NODELAY set
* Connected to ms.mdrinc.com (161.85.18.231) port 44321 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=IN; ST=Karnataka; L=Bangalore; O=mdrinc; OU=Dev; CN=*.mdrinc.com; emailAddress=joe@mdrinc.com
*  start date: Dec 12 16:20:14 2018 GMT
*  expire date: Dec 11 16:20:14 2023 GMT
*  common name: *.mdrinc.com (matched)
*  issuer: C=IN; ST=Karnataka; L=Bangalore; O=mdrinc; OU=Dev; CN=CA.HCPF.COM; emailAddress=a@b.com
*  SSL certificate verify ok.
> GET /api/values HTTP/1.1
> Host: ms.mdrinc.com:44321
> User-Agent: curl/7.58.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Thu, 13 Dec 2018 05:44:10 GMT
< Content-Type: application/json; charset=utf-8
< Transfer-Encoding: chunked
<
* Connection #0 to host ms.mdrinc.com left intact
["value1","value2"]
```

### My .NET Environment is
```
dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.2.100
 Commit:    b9f2fa0ca8

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/2.2.100/

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

.NET Core SDKs installed:
  2.2.100 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]_
```

### Example: my server returns following server certificate
```
[Version]
  V1

[Subject]
  E=joe@mdrinc.com, CN=*.mdrinc.com, OU=Dev, O=mdrinc, L=Bangalore, S=Karnataka, C=IN
  Simple Name: *.mdrinc.com
  Email Name: joe@mdrinc.com
  DNS Name: *.mdrinc.com

[Issuer]
  E=a@b.com, CN=CA.HCPF.COM, OU=Dev, O=mdrinc, L=Bangalore, S=Karnataka, C=IN
  Simple Name: CA.HCPF.COM
  Email Name: a@b.com
  DNS Name: CA.HCPF.COM

[Serial Number]
  009D69D9814A6C3377

[Not Before]
  12/12/18 9:50:14 PM

[Not After]
  11/12/23 9:50:14 PM

[Thumbprint]
  351A7F0941DA8B97D9773B391403521B30111327

[Signature Algorithm]
  sha256RSA(1.2.840.113549.1.1.11)

[Public Key]
  Algorithm: RSA
  Length: 2048
  Key Blob: 3082010A0282010100CE552505C5400505E42C8DD8D76899455EC86D8958A50BC3DEB4A0E09D8AD45AA7828EDA9CC3AF038A4B988818B07115881BD559CA0E39CCDC01D6D1C28F3A03F9E94FDD316FD77D0B8CB6390E7E5649A831D6201FC57114FCF7B84C88756DCCC0D1121E0D6A73D6D7D7206F90BD818D7CDB4A6B2D412A083763DA85CF83E6CE2DEC05AC226EE1828144F4DFB2B2721C8E71B584E1829B99FFCB6829671734D05D1A42AF1AFED72F9FD56C02E5516B3AC91D9707DC22B3B31E31DAFDC80D44FB96BE3136ACE54CFEB59A5802FCC867E5DADD5F6E3E59DAABA7EE5EEA2D7E8EBABA84A05F32B9E551A0D305B3A75BEB6F4ACFFD22E27AE7FC941463E361BFC4A30203010001
  Parameters: 0500
```

[WebAPI Sample](https://gist.github.com/madhub/25459c78560ca768fc44443d131dd643#file-webapi-cs)
[API Client ](https://gist.github.com/madhub/39c980ff010ba3888e54e8f1527ced6b#file-ssltest-cs)

### Sample Program
```csharp
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.Text;

class Program
{
    public class MyHttpClientHandler : HttpClientHandler
    {
        public MyHttpClientHandler()
        {
            this.ServerCertificateCustomValidationCallback = delegate (HttpRequestMessage message,
                X509Certificate2 x509Cer2,
                X509Chain chain,
                SslPolicyErrors errors)
            {
                bool validationStatus = true;
                if ((errors & SslPolicyErrors.RemoteCertificateChainErrors) != 0)
                {
                    Console.WriteLine("SslPolicyErrors.RemoteCertificateChainErrors");
                    validationStatus = false;
                }

                if ((errors & SslPolicyErrors.RemoteCertificateNameMismatch) != 0)
                {
                    Console.WriteLine("SslPolicyErrors.RemoteCertificateNameMismatch");
                    validationStatus = false;
                }
                else if ((errors & SslPolicyErrors.None) != 0)
                {
                    Console.WriteLine("SslPolicyErrors.None");
                    validationStatus = true;
                }

                Console.WriteLine(x509Cer2.ToString(true));

                return validationStatus;
            };
        }

    }
    // this works
    public static void WithAppContextSwitch(string hostUrl)
    {
        AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);
        try
        {
            using (var httpClient = new HttpClient())
            {
                var response = httpClient.GetAsync(hostUrl).GetAwaiter().GetResult();
                System.Console.WriteLine($"Http Response {response.StatusCode}");
                System.Console.WriteLine($"Response Content: {response.Content.ReadAsStringAsync().GetAwaiter().GetResult()}");
            }
        }
        catch (Exception exp)
        {

            Console.WriteLine("Exception " + exp);
        }
    }
  // throws RemoteCertificateNameMismatch exception
    public static void WithOutAppContextSwitch(string hostUrl)
    {

        try
        {
            var httpHandler = new MyHttpClientHandler();
            using (var httpClient = new HttpClient(httpHandler))
            {
                var response = httpClient.GetAsync(hostUrl).GetAwaiter().GetResult();
                System.Console.WriteLine($"Http Response {response.StatusCode}");
                System.Console.WriteLine($"Response Content: {response.Content.ReadAsStringAsync().GetAwaiter().GetResult()}");
            }
        }
        catch (Exception exp)
        {

            Console.WriteLine("Exception " + exp);
        }
    }

    static void Main(string[] args)
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            Console.WriteLine("Running on Windows");
        }else if ( RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
        {
            Console.WriteLine("Running on Linux");
        }
        // I have updated hosts file with "ms.mdrinc.com" mapping to ip
        string hostUrl = "https://ms.mdrinc.com:44321/api/values";

        // this works
        WithAppContextSwitch(hostUrl);

        // throws RemoteCertificateNameMismatch exception
        //WithOutAppContextSwitch(hostUrl);
    }
}
```

_Originally posted by @madhub in https://github.com/dotnet/corefx/issues/30388#issuecomment-446554623_

**Comment by madhub** [comment]
It looks like if SAN ( **subject alternative name** )   is missing in the certificate , wildcard matching is ignored in the .NET Core openssl implementation ("CryptoNative_CheckX509Hostname")

https://github.com/dotnet/corefx/blob/05d119eeb832fc0076ae4c94ddff57eb4909e88c/src/Native/Unix/System.Security.Cryptography.Native/openssl.c line 816

**Comment by bartonjs** [comment]
It doesn't look like your certificate has a subject alternative names extension.

.NET's hostname validator only matches wildcards within a dNSName label of the Subject Alternative Names extension.  The closest authority to this pattern that I can find is [RFC 4513]:

>    The server's identity may also be verified by comparing the reference
   identity to the Common Name (CN) [RFC4519] value in the leaf Relative
   Distinguished Name (RDN) of the subjectName field of the server's
   certificate.  This comparison is performed using the rules for
   comparison of DNS names in Section 3.1.3.1, below, **with the exception
   that no wildcard matching is allowed**.  Although the use of the Common
   Name value is existing practice, it is deprecated, and Certification
   Authorities are encouraged to provide subjectAltName values instead.

<sub>(emphasis added)</sub>

This matches Windows' validator, and is a practical match to public CA issued certificates, since CA/Browser Forum members all issue SAN extensions (and once a SAN extension is present the Subject CN is no longer used for a match).

**Comment by bartonjs** [comment]
@madhub Sorry, apparently I didn't hit the post button yesterday.

**Comment by madhub** [comment]
curl works properly, shown in example
I wrote java program to test it also works.
`openssl s_client -connect ms.mdrinc.com:44321 also works.`
Same program with `AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false)`; also works.
I am not sure what is right ?

**Comment by bartonjs** [comment]
`openssl s_client` never validates hostnames

There's not really any clear/crisp standard of what wildcard matching should be done.

* Curl is using the form of treating the CN value as an extra dNSName label, meaning wildcards are allowed.
* Java is either doing the same, or your test app has ended up disabling validation.
* .NET Core on Linux says that when falling back to the CN value, wildcards are not considered.
  * This is the proscribed behavior by RFC 4513, of course other RFCs may disagree.
  * This is the behavior that we were recommended to implement by the Windows teams responsible for certificates and TLS, so it should be the same as what HttpClient does on .NET Framework or .NET Core on Windows.

The [CA/Browser Forum's Baseline Requirements 1.6.2](https://cabforum.org/wp-content/uploads/CA-Browser-Forum-BR-1.6.2.pdf) says

> 7.1.4.2.1 Subject Alternative Name Extension
Certificate Field: extensions:subjectAltName
Required/Optional: **Required**
Contents: This extension MUST contain at least one entry. Each entry MUST be either a dNSName containing the Fully-Qualified Domain Name or an iPAddress containing the IP address of a server. The CA MUST confirm that the Applicant controls the Fully-Qualified Domain Name or IP address or has been granted the right to use it by the Domain Name Registrant or IP address assignee, as appropriate.
**Wildcard FQDNs are permitted.**

So a (conforming) CA is required to always issue Subject Alternative Name values.  Therefore "browsers" (of which HttpClient can be said to be one) can rely on it always being present.

While it's unfortunate that you've found a difference in the hostname validator when we let curl evaluate it (UseSocketsHttpHandler=false) vs when we do the HTTP ourselves on top of SslStream; expanding the match scope is something that would need a security signoff (since previously rejected connections would now be accepted) and not very likely to be implemented.

The easy workaround, of course, is to have your private PKI issue certificates with subject alternative names present.

**Comment by madhub** [comment]
Just to clarify `Java program` does not disable validation. I will post my Java sample  shortly.

Regarding the workaround, is this behavior is consistent on Windows & Linux version of .NET core ?, I am sure on Windows you are not using `openssl `library . 

I will do some more testing with other Http implementation `wget `/ `python HttpClient` /.

**Comment by bartonjs** [comment]
> is this behavior is consistent on Windows & Linux version of .NET core ?

My previous statement: "This is the behavior that we were recommended to implement by the Windows teams responsible for certificates and TLS, so it should be the same as what HttpClient does on .NET Framework or .NET Core on Windows."

Since I don't have a conforming setup at the moment (a no-SAN certificate with a wildcard CN issued by a trusted root), I can't verify that assertion; but evidence is welcome.

**Comment by madhub** [comment]
RFC 4513 you quoted is for LDAP , is the same rules apply for Http ?
I do not see anywhere it say's it is mandatory to fill subject alternative name. I tried to use digicert CSR utilities, it shows as optional. 
![image](https://user-images.githubusercontent.com/8907962/50045194-91b23780-00b4-11e9-992d-bf9dc6b3d422.png)
More I read about wildcard & SAN certs , they are for different purpose. If I have wildcard certs you should be  able to protect all **subdomains** [ wildcard & SAN certs](https://opensrs.com/blog/2012/09/san-and-wildcard-certificates-whats-the-difference/)

**Comment by madhub** [comment]
Hi 
Here is the update.
Sample  program attached in the issue **works on Windows** , here is the output both from Windows 10 and Linux

**Windows output.**
```
>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.2.100
 Commit:    b9f2fa0ca8

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.16299
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.100\

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

.NET Core SDKs installed:
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.2.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]


```
**Program Execution output on Windows**
```
dotnet run
Running on Windows
[Version]
  V1

[Subject]
  E=joe@mdrinc.com, CN=*.mdrinc.com, OU=Dev, O=mdrinc, L=Bangalore, S=Karnataka, C=IN
  Simple Name: *.mdrinc.com
  Email Name: joe@mdrinc.com
  DNS Name: *.mdrinc.com

[Issuer]
  E=a@b.com, CN=CA.HCPF.COM, OU=Dev, O=mdrinc, L=Bangalore, S=Karnataka, C=IN
  Simple Name: CA.HCPF.COM
  Email Name: a@b.com
  DNS Name: CA.HCPF.COM

[Serial Number]
  009D69D9814A6C3377

[Not Before]
  2018-12-12 9:50:14 PM

[Not After]
  2023-12-11 9:50:14 PM

[Thumbprint]
  351A7F0941DA8B97D9773B391403521B30111327

[Signature Algorithm]
  sha256RSA(1.2.840.113549.1.1.11)

[Public Key]
  Algorithm: RSA
  Length: 2048
  Key Blob: 30 82 01 0a 02 82 01 01 00 ce 55 25 05 c5 40 05 05 e4 2c 8d d8 d7 68 99 45 5e c8 6d 89 58 a5 0b c3 de b4 a0 e0 9d 8a d4 5a a7 82 8e da 9c c3 af 03 8a 4b 98 88 18 b0 71 15 88 1b d5 59 ca 0e 39 cc dc 01 d6 d1 c2 8f 3a 03 f9 e9 4f dd 31 6f d7 7d 0b 8c b6 39 0e 7e 56 49 a8 31 d6 20 1f c5 71 14 fc f7 b8 4c 88 75 6d cc c0 d1 12 1e 0d 6a 73 d6 d7 d7 20 6f 90 bd 81 8d 7c db 4a 6b 2d 41 2a 08 37 63 da 85 cf 83 e6 ce 2d ec 05 ac 22 6e e1 82 81 44 f4 df b2 b2 72 1c 8e 71 b5 84 e1 82 9b 99 ff cb 68 29 67 17 34 d0 5d 1a 42 af 1a fe d7 2f 9f d5 6c 02 e5 51 6b 3a c9 1d 97 07 dc 22 b3 b3 1e 31 da fd c8 0d 44 fb 96 be 31 36 ac e5 4c fe b5 9a 58 02 fc c8 67 e5 da dd 5f 6e 3e 59 da ab a7 ee 5e ea 2d 7e 8e ba ba 84 a0 5f 32 b9 e5 51 a0 d3 05 b3 a7 5b eb 6f 4a cf fd 22 e2 7a e7 fc 94 14 63 e3 61 bf c4 a3 02 03 01 00 01
  Parameters: 05 00

Http Response OK
Response Content: ["value1","value2"]

```

**Linux Output**

```
dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.2.100
 Commit:    b9f2fa0ca8

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/2.2.100/

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

.NET Core SDKs installed:
  2.2.100 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

```
**Program Execution output on Linux**
```
 dotnet run
Running on Linux
SslPolicyErrors.RemoteCertificateNameMismatch
[Version]
  V1

[Subject]
  E=joe@mdrinc.com, CN=*.mdrinc.com, OU=Dev, O=mdrinc, L=Bangalore, S=Karnataka, C=IN
  Simple Name: *.mdrinc.com
  Email Name: joe@mdrinc.com
  DNS Name: *.mdrinc.com

[Issuer]
  E=a@b.com, CN=CA.HCPF.COM, OU=Dev, O=mdrinc, L=Bangalore, S=Karnataka, C=IN
  Simple Name: CA.HCPF.COM
  Email Name: a@b.com
  DNS Name: CA.HCPF.COM

[Serial Number]
  009D69D9814A6C3377

[Not Before]
  12/12/18 9:50:14 PM

[Not After]
  11/12/23 9:50:14 PM

[Thumbprint]
  351A7F0941DA8B97D9773B391403521B30111327

[Signature Algorithm]
  sha256RSA(1.2.840.113549.1.1.11)

[Public Key]
  Algorithm: RSA
  Length: 2048
  Key Blob: 3082010A0282010100CE552505C5400505E42C8DD8D76899455EC86D8958A50BC3DEB4A0E09D8AD45AA7828EDA9CC3AF038A4B988818B07115881BD559CA0E39CCDC01D6D1C28F3A03F9E94FDD316FD77D0B8CB6390E7E5649A831D6201FC57114FCF7B84C88756DCCC0D1121E0D6A73D6D7D7206F90BD818D7CDB4A6B2D412A083763DA85CF83E6CE2DEC05AC226EE1828144F4DFB2B2721C8E71B584E1829B99FFCB6829671734D05D1A42AF1AFED72F9FD56C02E5516B3AC91D9707DC22B3B31E31DAFDC80D44FB96BE3136ACE54CFEB59A5802FCC867E5DADD5F6E3E59DAABA7EE5EEA2D7E8EBABA84A05F32B9E551A0D305B3A75BEB6F4ACFFD22E27AE7FC941463E361BFC4A30203010001
  Parameters: 0500

Exception System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception. ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.
   at System.Net.Security.SslState.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslStream.BeginAuthenticateAsClient(SslClientAuthenticationOptions sslClientAuthenticationOptions, CancellationToken cancellationToken, AsyncCallback asyncCallback, Object asyncState)
   at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__47_0(SslClientAuthenticationOptions arg1, CancellationToken arg2, AsyncCallback callback, Object state)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2](Func`5 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, Object state, TaskCreationOptions creationOptions)
   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2](Func`5 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, Object state, TaskCreationOptions creationOptions)
   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1,TArg2](Func`5 beginMethod, Action`1 endMethod, TArg1 arg1, TArg2 arg2, Object state)
   at System.Net.Security.SslStream.AuthenticateAsClientAsync(SslClientAuthenticationOptions sslClientAuthenticationOptions, CancellationToken cancellationToken)
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at DotNetCoreApp.WithOutAppContextSwitch(String hostUrl) in /home/madhub/MadhuAPIDemo/Program.cs:line 76

```

**Comment by madhub** [comment]
**It also works from the browser ( tested on Ubuntu Linux firefox browser)**
![image](https://user-images.githubusercontent.com/8907962/50073666-57a17c80-01ff-11e9-838a-ff4b6e68079b.png)

**Comment by madhub** [comment]
Any update on this issue ?
Is it going to be fixed ?

**Comment by madhub** [comment]
If you don't fix it , same program behaves  differently on Windows & Linux , is this acceptable ?.

**Comment by bartonjs** [comment]
When we initially released the minimum version of OpenSSL we used didn't hostname validation built in (it was up to the caller to write it).  Assuming that our older-OS matrix isn't a problem we'll move to use OpenSSL `X509_check_host` for the 3.0 release (removing our custom code).  Experimentation says that will fix the problem you're describing.

Note that while it will fix the problem in future releases, it won't fix 2.1 or 2.2.  If you want things to function on those releases you still need to fix your certificate creation process to produce a Subject Alternative Name extension if you want wildcards to work.

**Comment by madhub** [comment]
Thanks for providing the milestone for fixing the issue. 
Not fixing in 2.1 is a problem since 2.1 is LTS release , many organizations including ours relay LTS software for the deployment, please reconsider fixing it in 2.1 release.

I have one comment on your proposal for fixing the issue in  older release ( 2.1 or 2.2) .
I can change the **my self signed** certificate to include **Subject Alternative Name** ,  but larger issue is how do we handle similar issue in the software deployment where end user deploying our software is providing the certificate. We can't assume that all the certificates used in organizations are signed by the one of the global CA's who adhere to latest  "CA/Browser Forum's Baseline Requirements 1.6.2", that may not be the case ,many organizations use self signed certificates signed by their own self signed root certificate they might not have looked at the ""CA/Browser Forum's Baseline Requirements ". We cannot convince them to do the same as your described in your proposal,  as they **can say same certificate works for other services in java** , even on **.NET framework/.NET core on Windows.**

.NET core platform should handle these issue in tolerant way, similar to how browsers & other platform like java handle the issue.

More importantly the .NET Core API behavior should be consistent across implementation ( Windows/Linux/Mac) otherwise it is big problem for the developer relying on cross platform development.
